src/services/mapCode.ts

```diff
@@ -0,0 +1,365 @@
+                // range for replacement.
+                const matches = (scope as Block).statements.filter(stmt => nodes.some(node => matchNode(node, stmt)));
+                matches.sort((a, b) => a.pos - b.pos);
+                changeTracker.replaceNodeRangeWithNodes(file, matches[0], matches[matches.length - 1], nodes);
+                // はい、できました！
+                return;
+            }
+        }
+    }
+
+    // If we have no frame of reference, we'll just insert at the end of the
+    // nearest statements scope, or, if we can't find a block, at the top
+    // level of the source file.
+    let scopeStatements: NodeArray<Statement> = file.statements;
+    top: for (const locationGroup of focusLocations) {
+        for (const location of locationGroup) {
+            const block = findAncestor(getNodeAtPosition(file, location.start), isBlock);
+            if (block) {
+                scopeStatements = (block as Block).statements;
+                break top;
+            }
+        }
+    }
+    changeTracker.insertNodesAfter(file, scopeStatements[scopeStatements.length - 1], nodes);
+}
+
+function getNodeAtPosition(sourceFile: SourceFile, position: number): Node {
+    const isJavaScriptFile = isSourceFileJS(sourceFile);
+    let current: Node = sourceFile;
+    const getContainingChild = (child: Node) => {
+        if (child.pos <= position && (position < child.end || (position === child.end && (child.kind === SyntaxKind.EndOfFileToken)))) {
+            return child;
+        }
+    };
+    while (true) {
+        const child = isJavaScriptFile && hasJSDocNodes(current) && forEach(current.jsDoc, getContainingChild) || forEachChild(current, getContainingChild);
+        if (!child) {
+            return current;
+        }
+        current = child;
+    }
+}
+
+function getIdentifier(node: NamedDeclaration): string | undefined {
+    const name = node.name;
+    if (!name) return undefined;
+    switch (name.kind) {
+        case SyntaxKind.Identifier:
+            return name.text;
+        case SyntaxKind.PrivateIdentifier:
+            return name.escapedText as string;
+        case SyntaxKind.JsxNamespacedName:
+            return name.namespace.text + ":" + name.name.text;
+        default:
+            return name.getText();
+    }
+}
+
+function matchNode(a: Node, b: Node): boolean {
+    if (a.kind !== b.kind) {
+        return false;
+    }
+
+    if (a.kind === SyntaxKind.Constructor) {
+        return a.kind === b.kind;
+    }
+
+    if (isNamedDeclaration(a) && isNamedDeclaration(b)) {
+        const aName = getIdentifier(a);
+        const bName = getIdentifier(b);
+        return !!(aName && aName === bName);
+    }
+
+    if (a.getText() === b.getText()) {
+        return true;
+    }
+
+    // TODO: Maybe match by other characteristics?
+    return false;
+}
+
+function findAncestor(node: Node, f: (node: Node) => boolean): Node | undefined {

```

Comment:
I saw @andrewbranch point this out in another pr, there is an existing utility function `findAncestor()` in utilitiesPublic.ts that can be used here.